dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp

```diff
@@ -165,16 +314,159 @@ void DAGStorageInterpreter::execute(DAGPipeline & pipeline)
+        stream = std::make_shared<ExpressionBlockInputStream>(stream, project_after_where, log->identifier());
+    }
+}
 
-    null_stream_if_empty = std::make_shared<NullBlockInputStream>(storage_for_logical_table->getSampleBlockForColumns(required_columns));
+void DAGStorageInterpreter::executeCastAfterTableScan(
+    size_t remote_read_streams_start_index,
+    DAGPipeline & pipeline)
+{
+    auto original_source_columns = analyzer->getCurrentInputColumns();
 
-    // Should build these vars under protect of `table_structure_lock`.
-    buildRemoteRequests();
+    ExpressionActionsChain chain;
+    analyzer->initChain(chain, original_source_columns);
 
-    releaseAlterLocks();
+    // execute timezone cast or duration cast if needed for local table scan
+    if (addExtraCastsAfterTs(*analyzer, is_need_add_cast_column, chain, table_scan))
+    {
+        ExpressionActionsPtr extra_cast = chain.getLastActions();
+        chain.finalize();
+        chain.clear();
+
+        // After `addExtraCastsAfterTs`, analyzer->getCurrentInputColumns() has been modified.
+        // For remote read, `timezone cast and duration cast` had been pushed down, don't need to execute cast expressions.
+        // To keep the schema of local read streams and remote read streams the same, do project action for remote read streams.
+        NamesWithAliases project_for_remote_read;
+        const auto & after_cast_source_columns = analyzer->getCurrentInputColumns();
+        for (size_t i = 0; i < after_cast_source_columns.size(); ++i)
+        {
+            project_for_remote_read.emplace_back(original_source_columns[i].name, after_cast_source_columns[i].name);
+        }
+        assert(!project_for_remote_read.empty());
+        assert(pipeline.streams_with_non_joined_data.empty());
+        assert(remote_read_streams_start_index <= pipeline.streams.size());
+        size_t i = 0;
+        // local streams
+        while (i < remote_read_streams_start_index)
+        {
+            auto & stream = pipeline.streams[i++];
+            stream = std::make_shared<ExpressionBlockInputStream>(stream, extra_cast, log->identifier());
+        }
+        // remote streams
+        if (i < pipeline.streams.size())
+        {
+            ExpressionActionsPtr project_for_cop_read = generateProjectExpressionActions(
+                pipeline.streams[i],
+                context,
+                project_for_remote_read);
+            while (i < pipeline.streams.size())
+            {
+                auto & stream = pipeline.streams[i++];
+                stream = std::make_shared<ExpressionBlockInputStream>(stream, project_for_cop_read, log->identifier());
+            }
+        }
+    }
+}
+
+void DAGStorageInterpreter::executeRemoteQuery(DAGPipeline & pipeline)
+{
+    assert(!remote_requests.empty());
+    DAGSchema & schema = remote_requests[0].schema;
+#ifndef NDEBUG
+    auto schema_match = [&schema](const DAGSchema & other) {
```

Comment:
Why not move the schemaMatch function defined in DAGQueryBlockInterpreter into a util file? And reuse it.